## Preprocessing

In [3]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()


EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [6]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(['EIN', 'NAME'], axis=1)



In [7]:
# Determine the number of unique values in each column of application_df
application_df.nunique()

# Encode categorical variables using pd.get_dummies()
application_df_encoded = pd.get_dummies(application_df)

# Split the preprocessed data into feature and target arrays, then into training and testing datasets
y = application_df_encoded['IS_SUCCESSFUL'].values
X = application_df_encoded.drop('IS_SUCCESSFUL', axis=1).values

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Scale the training and testing data
scaler = StandardScaler()
X_scaler = scaler.fit(X_train)

X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


In [8]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)


APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [9]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff = 150  # Adjusted cutoff value

application_types_to_replace = list(application_type_counts[application_type_counts < cutoff].index)

print(application_types_to_replace)


# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

['T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']


APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
Other      120
Name: count, dtype: int64

In [10]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [11]:
# Look at CLASSIFICATION value counts that are greater than 1
classification_counts = application_df['CLASSIFICATION'].value_counts()
classification_counts_gt1 = classification_counts[classification_counts > 1]

print(classification_counts_gt1)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [12]:
# Choose a cutoff value for the classifications
cutoff = 100

# Create a list of classifications to be replaced
classifications_to_replace = list(classification_counts[classification_counts <= cutoff].index)

# Replace classifications in the DataFrame
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, "Other")

# Check to see if the replacement was successful
classification_counts_post_binning = application_df['CLASSIFICATION'].value_counts()
print(classification_counts_post_binning)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: count, dtype: int64


In [13]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)

# Display the first few rows of the new DataFrame with dummy variables
print(application_df_dummies.head())


   STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                   False   
1       1   108590              1                   False   
2       1     5000              0                   False   
3       1     6692              1                   False   
4       1   142590              1                   False   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                  True                 False                False   
1                 False                 False                 True   
2                 False                 False                False   
3                 False                 False                 True   
4                 False                 False                 True   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                False                False                False  ...   
1                False                False                False  

In [14]:
# Define the features set 'X' by selecting all columns except 'IS_SUCCESSFUL'
X = application_df_dummies.drop('IS_SUCCESSFUL', axis=1).values

# Define the target set 'y' by selecting 'IS_SUCCESSFUL'
y = application_df_dummies['IS_SUCCESSFUL'].values


from sklearn.model_selection import train_test_split

# Splitting the data into training and testing datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.2)


In [15]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [18]:
import tensorflow as tf

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
y_train = y_train.astype('int')  # or `.astype('float32')` for binary classification targets
y_test = y_test.astype('int')  # or `.astype('float32')`

import numpy as np

# Check for and handle NaNs
if np.isnan(X_train).any():
    X_train = np.nan_to_num(X_train)
if np.isnan(X_test).any():
    X_test = np.nan_to_num(X_test)

# Similarly for infinite values
if np.isinf(X_train).any():
    X_train = np.clip(X_train, a_min=-np.finfo(np.float32).max, a_max=np.finfo(np.float32).max)
if np.isinf(X_test).any():
    X_test = np.clip(X_test, a_min=-np.finfo(np.float32).max, a_max=np.finfo(np.float32).max)


# Determine the number of input features
n_features = X_train.shape[1]

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=n_features))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))  # 'sigmoid' for binary classification

# Check the structure of the model
nn.summary()

# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
fit_model = nn.fit(X_train, y_train, epochs=100)  # epochs is the number of iterations over the entire dataset

# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 80)                4080      
                                                                 
 dense_4 (Dense)             (None, 30)                2430      
                                                                 
 dense_5 (Dense)             (None, 1)                 31        
                                                                 
Total params: 6541 (25.55 KB)
Trainable params: 6541 (25.55 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100


858/858 [==============================] - 1s 765us/step - loss: 92850.6484 - accuracy: 0.4916
Epoch 2/100
858/858 [==============================] - 1s 743us/step - loss: 85028.8828 - accuracy: 0.5077
Epoch 3/100
858/858 [==============================] - 1s 729us/step

In [19]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - loss: 0.8176 - accuracy: 0.4983 - 138ms/epoch - 643us/step
Loss: 0.8175994157791138, Accuracy: 0.49825072288513184


In [20]:
# Save the model as an HDF5 file
nn.save("my_model.h5")


C:\Users\lisal\AppData\Roaming\Python\Python311\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [21]:
from tensorflow.keras.models import load_model

# Load the model
loaded_nn = load_model("my_model.h5")


In [22]:
# optimize the model
# Adjusting the model architecture
nn_optimized = tf.keras.models.Sequential()

# Input layer and first hidden layer
nn_optimized.add(tf.keras.layers.Dense(units=128, activation='relu', input_dim=n_features))

# Second hidden layer
nn_optimized.add(tf.keras.layers.Dense(units=64, activation='relu'))

# Maybe add another hidden layer
nn_optimized.add(tf.keras.layers.Dense(units=32, activation='relu'))

# Output layer
nn_optimized.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Compile the model
nn_optimized.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model with more epochs
fit_model = nn_optimized.fit(X_train, y_train, epochs=150, batch_size=10, validation_split=0.2)

# Evaluate model performance
model_loss, model_accuracy = nn_optimized.evaluate(X_test, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


Epoch 1/150
2196/2196 [==============================] - 3s 977us/step - loss: 27646.6992 - accuracy: 0.4826 - val_loss: 8615.3867 - val_accuracy: 0.5171
Epoch 2/150
2196/2196 [==============================] - 2s 916us/step - loss: 60233.4961 - accuracy: 0.4952 - val_loss: 190724.0156 - val_accuracy: 0.5171
Epoch 3/150
2196/2196 [==============================] - 2s 941us/step - loss: 24738.0273 - accuracy: 0.5079 - val_loss: 955.8054 - val_accuracy: 0.4829
Epoch 4/150
2196/2196 [==============================] - 2s 939us/step - loss: 13910.2344 - accuracy: 0.5049 - val_loss: 0.8722 - val_accuracy: 0.5171
Epoch 5/150
2196/2196 [==============================] - 2s 948us/step - loss: 0.7509 - accuracy: 0.5357 - val_loss: 0.7060 - val_accuracy: 0.5171
Epoch 6/150
2196/2196 [==============================] - 2s 955us/step - loss: 0.6924 - accuracy: 0.5357 - val_loss: 0.6938 - val_accuracy: 0.5171
Epoch 7/150
2196/2196 [==============================] - 2s 992us/step - loss: 0.6906 - accu

In [23]:
nn_adjusted = tf.keras.models.Sequential([
    tf.keras.layers.Dense(128, activation='relu', input_dim=n_features),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

nn_adjusted.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

fit_model = nn_adjusted.fit(X_train, y_train, epochs=150, batch_size=32, validation_data=(X_test, y_test))


Epoch 1/150
858/858 [==============================] - 2s 1ms/step - loss: 0.7108 - accuracy: 0.5133 - val_loss: 0.6916 - val_accuracy: 0.5341
Epoch 2/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6987 - accuracy: 0.5157 - val_loss: 0.6969 - val_accuracy: 0.5338
Epoch 3/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6942 - accuracy: 0.5206 - val_loss: 0.6913 - val_accuracy: 0.5345
Epoch 4/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6926 - accuracy: 0.5288 - val_loss: 0.6908 - val_accuracy: 0.5347
Epoch 5/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6923 - accuracy: 0.5280 - val_loss: 0.6910 - val_accuracy: 0.5341
Epoch 6/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6919 - accuracy: 0.5303 - val_loss: 0.6915 - val_accuracy: 0.5345
Epoch 7/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6918 - accuracy: 0.5287 - val_loss: 0.6911 - val_accuracy: 0.5341

In [24]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(128, activation='relu', input_shape=(n_features,)),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=150, validation_data=(X_test, y_test), batch_size=32)


Epoch 1/150
858/858 [==============================] - 2s 1ms/step - loss: 0.7370 - accuracy: 0.5070 - val_loss: 0.6974 - val_accuracy: 0.5341
Epoch 2/150
858/858 [==============================] - 1s 1ms/step - loss: 0.7039 - accuracy: 0.5237 - val_loss: 0.7045 - val_accuracy: 0.5341
Epoch 3/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6965 - accuracy: 0.5207 - val_loss: 0.6908 - val_accuracy: 0.5341
Epoch 4/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6942 - accuracy: 0.5301 - val_loss: 0.6933 - val_accuracy: 0.5341
Epoch 5/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6921 - accuracy: 0.5324 - val_loss: 0.6921 - val_accuracy: 0.5350
Epoch 6/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6919 - accuracy: 0.5320 - val_loss: 0.6941 - val_accuracy: 0.5341
Epoch 7/150
858/858 [==============================] - 1s 1ms/step - loss: 0.6919 - accuracy: 0.5319 - val_loss: 0.6913 - val_accuracy: 0.5350

In [25]:
from tensorflow.keras.callbacks import EarlyStopping

# Define early stopping callback
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

# Include the callback in the fit method
fit_model = nn.fit(
    X_train, y_train, epochs=150, 
    validation_data=(X_test, y_test),
    batch_size=32,
    callbacks=[early_stopping]
)


Epoch 1/150
858/858 [==============================] - 1s 936us/step - loss: 0.6913 - accuracy: 0.5320 - val_loss: 0.6908 - val_accuracy: 0.5341
Epoch 2/150
858/858 [==============================] - 1s 914us/step - loss: 0.6913 - accuracy: 0.5320 - val_loss: 0.6908 - val_accuracy: 0.5341
Epoch 3/150
858/858 [==============================] - 1s 897us/step - loss: 0.6912 - accuracy: 0.5320 - val_loss: 0.6908 - val_accuracy: 0.5341
Epoch 4/150
858/858 [==============================] - 1s 910us/step - loss: 0.6913 - accuracy: 0.5320 - val_loss: 0.6908 - val_accuracy: 0.5341
Epoch 5/150
858/858 [==============================] - 1s 916us/step - loss: 0.6912 - accuracy: 0.5320 - val_loss: 0.6911 - val_accuracy: 0.5341
Epoch 6/150
858/858 [==============================] - 1s 939us/step - loss: 0.6914 - accuracy: 0.5320 - val_loss: 0.6909 - val_accuracy: 0.5341
Epoch 7/150
858/858 [==============================] - 1s 904us/step - loss: 0.6913 - accuracy: 0.5320 - val_loss: 0.6908 - val_ac

In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Initialize the RandomForest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the model
rf_model.fit(X_train, y_train)

# Predictions
y_pred = rf_model.predict(X_test)

# Evaluate the model
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")


Accuracy: 0.7131195335276967


In [27]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Define the parameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize the classifier
rf = RandomForestClassifier(random_state=42)

# Initialize GridSearchCV
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2, scoring='accuracy')

# Fit GridSearchCV
grid_search.fit(X_train, y_train)

# Best parameters and best score
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

# Evaluate on the test set
best_rf = grid_search.best_estimator_
y_pred = best_rf.predict(X_test)
print(f"Test Accuracy: {accuracy_score(y_test, y_pred)}")


Fitting 3 folds for each of 108 candidates, totalling 324 fits
Best Parameters: {'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 200}
Best Score: 0.7322788697693247
Test Accuracy: 0.7287172011661808


In [28]:
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

# Define the base models
model1 = LogisticRegression(random_state=42)
model2 = RandomForestClassifier(n_estimators=100, random_state=42)
model3 = DecisionTreeClassifier(random_state=42)

# Create a voting classifier
voting_clf = VotingClassifier(estimators=[('lr', model1), ('rf', model2), ('dt', model3)], voting='hard')

# Fit the model
voting_clf.fit(X_train, y_train)

# Evaluate the model
y_pred = voting_clf.predict(X_test)
print(f"Voting Classifier Accuracy: {accuracy_score(y_test, y_pred)}")


Voting Classifier Accuracy: 0.7049562682215743


In [29]:
from sklearn.preprocessing import StandardScaler

# Initialize the scaler
scaler = StandardScaler()

# Fit on training data
X_train_scaled = scaler.fit_transform(X_train)

# Transform test data
X_test_scaled = scaler.transform(X_test)


In [31]:
# Calculate the accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")


Accuracy: 0.7049562682215743


In [33]:
import pandas as pd
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [38]:
import requests

url = "https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv"
r = requests.get(url, allow_redirects=True)

# Open the file and save the content of the request
with open("charity_data.csv", 'wb') as file:
    file.write(r.content)


In [39]:
df = pd.read_csv("charity_data.csv")


In [40]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(['EIN', 'NAME'], axis=1)

# Convert categorical data to numeric with `pd.get_dummies`
application_df_encoded = pd.get_dummies(application_df, drop_first=True)

# Split my preprocessed data into my features and target arrays
X = application_df_encoded.drop('IS_SUCCESSFUL', axis=1).values
y = application_df_encoded['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [41]:
from sklearn.preprocessing import StandardScaler

# Initialize the scaler
scaler = StandardScaler()

# Fit on training data and transform
X_train_scaled = scaler.fit_transform(X_train)

# Transform test data
X_test_scaled = scaler.transform(X_test)


In [42]:
import tensorflow as tf

# Define the model - deep neural net
model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(80, activation='relu', input_dim=X_train_scaled.shape[1]),
    tf.keras.layers.Dense(30, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
fit_model = model.fit(X_train_scaled, y_train, epochs=100, validation_split=0.2)


Epoch 1/100
686/686 [==============================] - 1s 1ms/step - loss: 0.5845 - accuracy: 0.7114 - val_loss: 0.5553 - val_accuracy: 0.7376
Epoch 2/100
686/686 [==============================] - 1s 971us/step - loss: 0.5570 - accuracy: 0.7254 - val_loss: 0.5542 - val_accuracy: 0.7387
Epoch 3/100
686/686 [==============================] - 1s 939us/step - loss: 0.5527 - accuracy: 0.7280 - val_loss: 0.5530 - val_accuracy: 0.7403
Epoch 4/100
686/686 [==============================] - 1s 927us/step - loss: 0.5500 - accuracy: 0.7285 - val_loss: 0.5512 - val_accuracy: 0.7392
Epoch 5/100
686/686 [==============================] - 1s 925us/step - loss: 0.5496 - accuracy: 0.7289 - val_loss: 0.5502 - val_accuracy: 0.7391
Epoch 6/100
686/686 [==============================] - 1s 964us/step - loss: 0.5473 - accuracy: 0.7307 - val_loss: 0.5489 - val_accuracy: 0.7418
Epoch 7/100
686/686 [==============================] - 1s 939us/step - loss: 0.5473 - accuracy: 0.7312 - val_loss: 0.5504 - val_accu

In [43]:
# Evaluate the model using the test data
model_loss, model_accuracy = model.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


215/215 - 0s - loss: 0.5622 - accuracy: 0.7280 - 142ms/epoch - 660us/step
Loss: 0.5622308254241943, Accuracy: 0.7279883623123169


In [44]:
# Save the model
model.save("AlphabetSoupCharity_Optimization.h5")


C:\Users\lisal\AppData\Roaming\Python\Python311\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


Report on the Neural Network Model for Alphabet Soup Charity Prediction
Overview of the Analysis
The purpose of this analysis was to create a machine learning model capable of predicting whether applicants to Alphabet Soup’s charity funding would be successful in using the funds effectively. Using features extracted from historical data, a deep learning model was developed and optimized with the goal of achieving a target predictive accuracy higher than 75%.

Results
Data Preprocessing
Target Variable: The target for the model was IS_SUCCESSFUL, indicating whether the funding was used effectively.
Features for the Model: The model features included variables such as APPLICATION_TYPE, AFFILIATION, CLASSIFICATION, USE_CASE, ORGANIZATION, STATUS, INCOME_AMT, SPECIAL_CONSIDERATIONS, and ASK_AMT, which represent various characteristics of the applications.
Variables Removed: The variables EIN and NAME were removed from the input data as they are identification information and do not contribute to the model's ability to predict successful use of funds.
Compiling, Training, and Evaluating the Model
Model Architecture: The neural network model consisted of an input layer, two hidden layers, and an output layer. The first hidden layer had 80 neurons, and the second hidden layer had 30 neurons, both using the ReLU activation function. The output layer used a sigmoid activation function for binary classification.
Achieving Target Performance: The final model achieved an accuracy of approximately 73.8% on the validation set, which is below the target of 75%.
Optimization Attempts: To increase model performance, several strategies were employed, including adjusting the number of neurons in the hidden layers, experimenting with different activation functions, and increasing the number of training epochs. The data was also preprocessed by scaling numerical features and encoding categorical variables.
Summary
The deep learning model developed for Alphabet Soup Charity prediction achieved a maximum accuracy of approximately 73.8%, falling slightly short of the 75% target. Despite various optimization attempts, including adjusting the model architecture and preprocessing steps, the target performance was not reached.

Recommendation for Different Model
Given the challenges in achieving the target accuracy with the current neural network model, it is recommended to explore ensemble machine learning models, such as Gradient Boosting Machines (GBM) or a Random Forest Classifier, for this classification problem. These models can handle a mix of numerical and categorical data well and often provide high accuracy and robustness against overfitting with proper tuning.

Why a Different Model Could Be Better: Ensemble models like GBM or Random Forest work by combining multiple weak learners to create a strong predictive model. They are often more interpretable and can provide insights into feature importance, helping to understand what drives successful funding applications. Additionally, these models can be less sensitive to noisy data and may perform better with the given dataset after appropriate hyperparameter tuning.
